src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs

```diff
@@ -0,0 +1,474 @@
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "0 0");
+        }
+
+        [Fact]
+        public void Generic()
+        {
+            string source = @"
+class C1<T>
+{
+    static void M()
+    {
+        T t = default;
+    }
+}
+class C2<T> where T : struct
+{
+    static void M()
+    {
+        T t = default;
+    }
+}
+";
+            // PROTOTYPE(default) The constrained T should be ok to infer "default" from
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            comp.VerifyDiagnostics(
+                // (13,14): error CS8300: Cannot convert default to 'T' because it is a value or nullable type
+                //        T t = default;
+                Diagnostic(ErrorCode.ERR_RefCantBeDefault, "default").WithArguments("T").WithLocation(13, 14),
+                // (6,14): error CS8300: Cannot convert default to 'T' because it is a value or nullable type
+                //        T t = default;
+                Diagnostic(ErrorCode.ERR_RefCantBeDefault, "default").WithArguments("T").WithLocation(6, 14)
+                );
+        }
+
+        [Fact]
+        public void DynamicInvocation()
+        {
+            string source = @"
+class C
+{
+    static void M1()
+    {
+        dynamic d = null;
+        d.M2(default);
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            comp.VerifyDiagnostics();
+            // PROTOTYPE(default) Crash
+        }
+
+        [Fact]
+        public void IndexingIntoArray()
+        {
+            string source = @"
+class C
+{
+    static void M1()
+    {
+        int[] x = { 1, 2 };
+        System.Console.Write(x[default]);
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions);
+            comp.VerifyDiagnostics(
+                // (7,32): error CS8301: Cannot infer the type of implicitly-typed default.
+                //         System.Console.Write(x[default]);
+                Diagnostic(ErrorCode.ERR_TypeInferenceFailedForImplicitlyTypedDefault, "default").WithLocation(7, 32)
+                );
+        }
+
+        [Fact]
+        public void Lambda()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        System.Console.Write(M()());
+    }
+    static System.Func<int> M()
+    {
+        return () => default;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.ExperimentalParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "0");
+        }
+    }
+}
```

Comment:
Other cases to test (that should work)

`switch (2) { case default: ; }`
`if (2 != default) ...`
`var a = new[] { 0, default };`
`void M(int x = default);`
`var a = new int[default];`